WOR-264 De-emphasize Other Costs

msilva-broad opened the pull request:
- Billing report now shows "other costs" in a footnote, and removes the card, leaving only 3 cost cards in summary area.
- updated end2end integration test.
- visual testing done locally.
---

aria-live region fixes
- screen reader announcement on report render is now consistent with content being announced in order.
- changed over to a single larger aria-live region for the entire report content, which is marked with aria-busy when loading to cut down on screen-reader noise.
- promoted a couple of sub-controls to be full React.FC components instead of just rendering sub-routines.  This looks to enable React rendering optimizations better, and help a great deal with removing some screen reader repetition of the entire live region content for a single report refresh.
- KNOWN ISSUE: still seeing initial announcement of the cost card headers place-holder content on initial load and hard tab switches (like switching projects.  I suspect this may be related to aria-controls prop not being implemented in our Tabs control/usage.  I will file a seperate ticket for that fix.
---
![image](https://user-images.githubusercontent.com/108079391/179776847-a6695f8c-feaf-4c45-add5-c43ca0f3e98f.png)

msilva-broad committed:
WOR-264 De-emphasize Other Costs
- Billing report now shows "other costs" in a footnote, and removes the card, leaving only 3 cost cards in summary area.
- updated end2end integration test.

cahrens commented:
When I try this out with VoiceOver, the aria-live region for the "other cost" explanation is getting read before the cards for total, compute, and storage. It would be great if that could be reversed… not sure what is causing it.

msilva-broad committed:
WOR-264 De-emphasize Other Costs
- changed "other costs" footnote to use aria-describedby pattern instead of aria-live for hopefully better ordering on a11y screen readers

cahrens commented on src/pages/billing/Project.js:
```diff
@@ -195,6 +195,7 @@ const groupByBillingAccountStatus = (billingProject, workspaces) => {
 const LazyChart = lazy(() => import('src/components/Chart'))
 const maxWorkspacesInChart = 10
 const spendReportKey = 'spend report'
+const otherCostsDomId = 'billing-other-costs'
```
This approach will work, and we should never have multiple of these sections on the page. However, the general pattern in the repo is to use `IdContainer` to generate the ID and obtain the reference to it. For example, on line 602.

msilva-broad replied:
removed from newer implementation

cahrens commented on src/pages/billing/Project.js:
```diff
@@ -195,6 +195,8 @@ const groupByBillingAccountStatus = (billingProject, workspaces) => {
 const spendReportKey = 'spend report'
+const otherCostsDomId = 'billing-other-costs'
+const otherMessaging = cost => `Total spend includes ${cost} in other infrastructure or query costs related to the general operations of Terra.`
 
```
The message looks pretty odd when the cost is "…" during the loading time. I get that we probably want to have some message to avoid text elements shifting once the cost has been loaded, but can we possibly have something else that reads better? Perhaps just "Total spend includes infrastructure or query costs related to the general operations of Terra"?

msilva-broad replied:
fixed

msilva-broad committed:
WOR-264 aria-live region fixes
- screen reader announcement on report render is now consistent with content being announced in order.
- changed over to a single larger aria-live region for the entire report content, which is marked with aria-busy when loading to cut down on screen-reader noise.
- promoted a couple of sub-controls to be full React.FC components instead of just rendering sub-routines.  This looks to enable React rendering optimizations better, and help a great deal with removing some screen reader repetition of the entire live region content for a single report refresh.
- KNOWN ISSUE: still seeing initial announcement of the cost card headers place-holder content on initial load and hard tab switches (like switching projects.  I suspect this may be related to aria-controls prop not being implemented in our Tabs control/usage.  I will file a seperate ticket for that fix.

msilva-broad committed:
- demoted cost cards headers to just divs
- removed report chart from live-region
- KNOWN ISSUE: React DOM update optimizations still causing screen reader inconsistencies and issues.  Exploring alternatives.

msilva-broad committed:
- reverting to general layout and aria-live strategy previously in place.
- implementing "other costs messaging" block to use same aria-live and aria=atomic behavior as cost cards.
KNOWN ISSUE: aria-live announcement is somewhat chatty and out-of-order depending on screen reader.  Leaving in place until more detailed investigation on React friendly aria-live technique can be researched.

msilva-broad committed:
- minor layout spacing fixes

cahrens commented on src/pages/billing/Project.js:
```diff
@@ -393,41 +401,21 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
-          ),
-          h(OtherMessaging, { cost: isProjectCostReady ? projectCost['other'] : null })
+          ]),
+          ...(_.map(name => CostCard({
```
@msilva-broad I thought there were good reasons to keep CostCard as a component. Does it cause issues as a component?

msilva-broad replied:
I thought so too, but as I reverted to the original setup, I was still seeing some wierdness and was tired of fussing with it all... reverting fixed things.  so, meh.

cahrens committed:
Fix bug where selecting an already-selected date range clears the cost displays.

cahrens commented on src/pages/billing/Project.js:
```diff
@@ -278,21 +288,39 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
+      div(
+        {
+          style: { flex: 'none', padding: '0.625rem 1.25rem' },
+          'aria-live': projectCost !== null ? 'polite' : 'off',
```
Nit: for consistency you could replace `projectCost !== null` with `isProjectCostReady`.

cahrens replied:
I pushed this up.

cahrens commented on src/pages/billing/Project.js:
```diff
@@ -278,21 +288,39 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
+          div({ style: { fontSize: 16, color: colors.accent(), margin: '0.25rem 0.0rem', fontWeight: 'normal' } }, [title]),
+          div({ style: { fontSize: 32, height: 40, fontWeight: 'bold', gridRowStart: '2' } }, [
+            amount,
+            (!!showAsterisk && isProjectCostReady) ? span(
```
Nit: `showAsterisk` is a boolean, so I think having the `!!` in front just makes it harder to read.

cahrens commented on src/pages/billing/Project.js:
```diff
@@ -361,19 +396,32 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
                 onChange: ({ value: selectedDays }) => {
-                  setSpendReportLengthInDays(selectedDays)
-                  setProjectCost(null)
+                  if (selectedDays !== spendReportLengthInDays) {
```
@msilva-broad This is a pre-existing bug that was bothering me so I pushed up a commit to fix it. Selecting an already-selected date range caused the cost cards to all clear and never reload.

cahrens committed:
Reference cost card as a component.

cahrens committed:
Nit.

cahrens committed:
Move CostCard component outside of render.
